File: ofdrw-java/ofdrw-layout/src/main/java/org/ofdrw/layout/engine/render/CanvasRender.java
package org.ofdrw.layout.engine.render;

import org.ofdrw.core.basicStructure.pageObj.layer.block.CT_PageBlock;
import org.ofdrw.core.basicType.ST_Box;
import org.ofdrw.core.basicType.ST_Loc;
import org.ofdrw.layout.element.Div;
import org.ofdrw.layout.element.canvas.Canvas;
import org.ofdrw.layout.element.canvas.DrawContext;
import org.ofdrw.layout.element.canvas.Drawer;
import org.ofdrw.layout.engine.ResManager;

import java.io.IOException;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Canvas 渲染器
 * <p>
 * {@link org.ofdrw.layout.element.canvas.Canvas} 的渲染器
 *
 * @author 权观宇
 * @since 2020-05-01 17:44:41
 */
public class CanvasRender implements Processor {


    /**
     * 执行Canvas 渲染器
     *
     * @param layer      将要渲染到的图层
     * @param resManager 资源管理器
     * @param canvas     canvas对象
     * @param maxUnitID  最大ID
     * @throws RenderException 渲染发生错误
     */
    public static void render(CT_PageBlock layer,
                              ResManager resManager,
                              Canvas canvas,
                              AtomicInteger maxUnitID) throws RenderException {
        // 获取绘制器
        Drawer drawer = canvas.getDrawer();
        if (drawer == null) {
            return;
        }
        // 尝试获取Canvas中的页块
        CT_PageBlock block = canvas.getPreferBlock();
        if (block == null) {
            // 若页块为空，则创建新的页块
            block = new CT_PageBlock();
            block.setObjID(maxUnitID.incrementAndGet());
            canvas.setPreferBlock(block);
            // 添加到图层中
            layer.addPageBlock(block);
        }


        // 根据盒式模型计算出加上 外边距、边框、内边距
        ST_Box boundary = new ST_Box(
                canvas.getX() + canvas.getMarginLeft() + canvas.getBorderLeft() + canvas.getPaddingLeft(),
                canvas.getY() + canvas.getMarginTop() + canvas.getBorderTop() + canvas.getPaddingTop(),
                canvas.getWidth(),
                canvas.getHeight());
        // 构建上下文
        try (DrawContext ctx = new DrawContext(block, boundary, maxUnitID, resManager)) {
            // 执行绘制工作
            drawer.draw(ctx);
        } catch (IOException e) {
            throw new RenderException("Canvas绘制过程中发生异常", e);
        }
    }

    /**
     * 渲染Canvas对象
     *
     * @param pageLoc    页面在虚拟容器中绝对路径。
     * @param layer      图片将要放置的图层
     * @param resManager 资源管理器
     * @param e          Canvas对象
     * @param maxUnitID  最大元素ID提供器
     * @throws RenderException 渲染发生错误
     */
    @Override
    public void render(ST_Loc pageLoc, CT_PageBlock layer, ResManager resManager, Div e, AtomicInteger maxUnitID)throws RenderException {
        if (e instanceof Canvas) {
            render(layer, resManager, (Canvas) e, maxUnitID);
        }
    }
}

File: OfdrwNet.Layout/Engine/Render/CanvasRender.cs
using System;
using System.IO;
using OfdrwNet.Core.BasicStructure.PageObj.Layer.Block;
using OfdrwNet.Core.BasicType;
using OfdrwNet.Layout.Element;
using OfdrwNet.Layout.Element.Canvas;
using OfdrwNet.Layout.Engine;

namespace OfdrwNet.Layout.Engine.Render
{
    /// <summary>
    /// Canvas渲染器
    /// <para>
    /// Canvas 的渲染器
    /// </para>
    /// 
    /// 作者: 权观宇
    /// 起始时间: 2020-05-01 11:35:00
    /// </summary>
    public class CanvasRender : IProcessor
    {
        /// <summary>
        /// 执行Canvas元素渲染
        /// </summary>
        /// <param name="pageLoc">页面在虚拟容器中绝对路径。</param>
        /// <param name="layer">图层</param>
        /// <param name="resManager">资源管理器</param>
        /// <param name="e">元素</param>
        /// <param name="maxUnitIDProvider">最大ID提供器</param>
        /// <exception cref="RenderException">渲染发生错误</exception>
        public void Render(StLoc pageLoc, CtPageBlock layer, ResManager resManager, Div e, Func<int> maxUnitIDProvider)
        {
            if (e is Canvas canvas)
            {
                Render(layer, resManager, canvas, maxUnitIDProvider);
            }
        }

        /// <summary>
        /// 渲染Canvas元素到指定图层
        /// </summary>
        /// <param name="layer">图层</param>
        /// <param name="resManager">资源管理器</param>
        /// <param name="canvas">Canvas对象</param>
        /// <param name="maxUnitIDProvider">对象ID提供者</param>
        public static void Render(CtPageBlock layer, ResManager resManager, Canvas canvas, Func<int> maxUnitIDProvider)
        {
            if (canvas == null)
            {
                return;
            }

            // 首先渲染Div的基本样式（背景、边框等）
            DivRender.Render(layer, canvas, maxUnitIDProvider);

            // 获取绘制器
            var drawer = canvas.GetDrawer();
            if (drawer == null)
            {
                return;
            }

            // 计算Canvas的实际绘制区域
            var x = canvas.GetX() + canvas.GetMarginLeft() + canvas.GetBorderLeft() + canvas.GetPaddingLeft();
            var y = canvas.GetY() + canvas.GetMarginTop() + canvas.GetBorderTop() + canvas.GetPaddingTop();
            var width = canvas.GetWidth();
            var height = canvas.GetHeight() ?? 0;

            // 创建页面块用于容纳Canvas绘制内容
            CtPageBlock canvasBlock;
            if (canvas.GetPreferBlock() != null)
            {
                canvasBlock = canvas.GetPreferBlock();
            }
            else
            {
                canvasBlock = new CtPageBlock(new StId(maxUnitIDProvider()));
                canvas.SetPreferBlock(canvasBlock);
                layer.Add(canvasBlock);
            }

            // 创建绘制上下文
            var boundary = new StBox(x, y, width, height);
            var drawContext = new DrawContext(canvasBlock, boundary, maxUnitIDProvider, resManager);

            try
            {
                // 执行绘制
                drawer.Draw(drawContext);
            }
            catch (Exception ex)
            {
                throw new RenderException($"Canvas绘制过程中发生错误: {ex.Message}", ex);
            }
            finally
            {
                // 清理绘制上下文
                drawContext?.Dispose();
            }
        }
    }
}